### ISSUE TITLE
Overly aggressive purging of get_users_in_room cache?

### ISSUE THREAD

**Issue by clokep** [matrix-org/synapse#11875]
We invalidate the `get_users_in_room` cache (and a few others) as part of [`_invalidate_state_caches`](https://github.com/matrix-org/synapse/blob/5305a5e88144828419249fd9e4c5198d92276a44/synapse/storage/_base.py#L59-L75), we [calculate a `members_changed`](https://github.com/matrix-org/synapse/blob/2aa37a4250675f6d9feb57ec0dce65b2a6a3cde6/synapse/storage/databases/main/events.py#L1105-L1120) with the comment:

> Figure out the changes of membership to invalidate the `get_rooms_for_user` cache.

But don't seem to be using it in the opposite direction (`get_users_for_room`).

tl;dr I think we are invalidating this cache in situations we do not need to.
